libflux/src/semantic/types.rs

```diff
@@ -141,16 +369,109 @@ impl Fresher {
     }
 }
 
+impl Tvar {
+    fn apply(self, sub: &Subst) -> MonoType {
+        match sub.lookup(self) {
+            Some(t) => t.clone(),
+            None => MonoType::Var(self),
+        }
+    }
+
+    fn unify(self, with: MonoType, cons: Const) -> Result {
+        match with {
+            MonoType::Var(tv) => {
+                if self == tv {
+                    // The empty substitution will always
+                    // unify a type variable with itself.
+                    Ok((Subst::empty(), cons))
+                } else {
+                    // Unify two distinct type variables.
+                    // This will update the kind constraints
+                    // associated with these type variables.
+                    self.unify_with_tvar(tv, cons)
+                }
+            }
+            _ => {
+                if with.contains(self) {
+                    // Invalid recursive type
+                    Err(Error::occurs_check(self, with))
+                } else {
+                    // Unify a type variable with a monotype.
+                    // The monotype must satisify any
+                    // constraints placed on the type variable.
+                    self.unify_with_type(with, cons)
+                }
+            }
+        }
+    }
+
+    fn unify_with_tvar(self, tv: Tvar, cons: Const) -> Result {
+        let mut cons = cons;
+        // Gather the kind constraints for both type variables
+        let kinds: HashSet<Kind> = cons
+            .0
+            .get(&self)
+            .unwrap_or(&HashSet::new())
+            .union(cons.0.get(&tv).unwrap_or(&HashSet::new()))
+            .map(|&kind| kind)
+            .collect();
+        let sub = Subst::init(maplit::hashmap! {self => MonoType::Var(tv)});
+        if kinds.len() > 0 {
+            // Update the kind constraints
+            cons.0.remove(&self);
+            cons.0.insert(tv, kinds);
+        };
+        Ok((sub, cons))
+    }
+
+    fn unify_with_type(self, t: MonoType, cons: Const) -> Result {
+        let mut cons = cons;
+        let sub = Subst::init(maplit::hashmap! {self => t.clone()});
+        if let Some(kinds) = cons.0.remove(&self) {
+            // The monotype must satisfy all the constraints of
+            // the type variable.
+            let (s, cons) = t.constrain(kinds, cons)?;
+            Ok((sub.merge(s), cons))
+        } else {
+            Ok((sub, cons))
+        }
+    }
+
+    fn constrain(self, with: HashSet<Kind>, cons: Const) -> Result {
```

Comment:
Ok, I see. You need `Const` above because you need it here, right?

But there is no trait to implement, so... 🤔

Reply:
It's also needed for Array, Row, and Function.